src/pip/_internal/index.py

```diff
@@ -270,16 +271,44 @@ class FoundCandidates(object):
     def __init__(
         self,
         candidates,     # type: List[InstallationCandidate]
-        specifier,      # type: specifiers.BaseSpecifier
-        prereleases,    # type: Optional[bool]
+        versions,       # type: Set[_BaseVersion]
```

Comment:
Since we're only ever checking whether strings belong to this set, is there a way we can use `str` rather than `_BaseVersion`? That would make the API simpler because I'm not sure what implications `_BaseVersion` has.

Reply:
I made `self._versions` a `Set[str]` instead. I also move class method back to `__init__` since it does not really make sense to allow creating a FoundCandidates with arbitrary version strings.

Reply:
I actually liked the approach you took before of having a creation class method. One reason is that it keeps the `__init__()` method "dumb" (without non-trivial code). Another is that it simplifies testing (e.g. you don't need to go through the indirection of creating a `BaseSpecifier` to test the logic of the class). Regarding your concern, you can just say in the docstring of the `__init__` method that the class method should be used (in other words that it's a private constructor).

Reply:
Sounds reasonable, thanks, I’ll bring back the class method then.